File: 24_高级数据结构/珂朵莉树/VanEmdeBoasTree.go
// Van Emde Boas Tree (梵峨眉大悲寺树)

package main

import (
	"bufio"
	"fmt"
	"os"
	"time"
)

func main() {
	demo := func() {
		van := NewVanEmdeBoasTree()
		fmt.Println(van.Min(), van.Max(), van.Size(), van.Prev(0), van.Next(0))
		van.Insert(1)
		van.Insert(2)
		fmt.Println(van.Has(1), van.Has(2), van.Has(3))
		fmt.Println(van.Prev(1), van.Prev(2), van.Prev(3), van.Prev(100), van.Prev(-1000000000000))
		van.Erase(1)
		fmt.Println(van.Min(), van.Max(), van.Size())
		van.Insert(-111)
		fmt.Println(van.Min(), van.Max(), van.Size())
		fmt.Println(van.Has(-111), van.Prev(-1), van.Min())

		n := int(5e5)
		fs := NewVanEmdeBoasTree()
		time1 := time.Now()
		for i := 0; i < n; i++ {
			fs.Insert(i)
			fs.Next(i)
			fs.Prev(i)
			fs.Has(i)
			fs.Erase(i)
			fs.Insert(i)
		}
		fmt.Println(time.Since(time1)) // !5e5 => 234ms(depth=8)
	}
	_ = demo
	// demo()

	// https://judge.yosupo.jp/problem/predecessor_problem
	in := bufio.NewReader(os.Stdin)
	out := bufio.NewWriter(os.Stdout)
	defer out.Flush()

	var n, q int
	fmt.Fscan(in, &n, &q)
	set := NewVanEmdeBoasTree()
	var s string
	fmt.Fscan(in, &s)
	for i, v := range s {
		if v == '1' {
			set.Insert(i)
		}
	}

	for i := 0; i < q; i++ {
		var op int
		fmt.Fscan(in, &op)
		switch op {
		case 0:
			var k int
			fmt.Fscan(in, &k)
			set.Insert(k)
		case 1:
			var k int
			fmt.Fscan(in, &k)
			set.Erase(k)
		case 2:
			var k int
			fmt.Fscan(in, &k)
			if set.Has(k) {
				fmt.Fprintln(out, 1)
			} else {
				fmt.Fprintln(out, 0)
			}
		case 3:
			var k int
			fmt.Fscan(in, &k)
			ceiling := set.Next(k)
			if ceiling != INF {
				fmt.Fprintln(out, ceiling)
			} else {
				fmt.Fprintln(out, -1)
			}
		case 4:
			var k int
			fmt.Fscan(in, &k)
			floor := set.Prev(k)
			if floor != -INF {
				fmt.Fprintln(out, floor)
			} else {
				fmt.Fprintln(out, -1)
			}

		}
	}
}

const INF int = 1e18

type VanEmdeBoasTree struct {
	root *VNode
	size int
}

// !建立一个元素范围为(-INF,INF)的VanEmdeBoasTree.
func NewVanEmdeBoasTree() *VanEmdeBoasTree {
	return &VanEmdeBoasTree{root: NewVNode(32)} // 16/32/64
}

func (van *VanEmdeBoasTree) Has(x int) bool {
	return van.root.Has(x)
}

func (van *VanEmdeBoasTree) Insert(x int) bool {
	if van.Has(x) {
		return false
	}
	van.size++
	van.root.Insert(x)
	return true
}

func (van *VanEmdeBoasTree) Erase(x int) bool {
	if !van.Has(x) {
		return false
	}
	van.size--
	van.root.Erase(x)
	return true
}

// 返回小于等于i的最大元素.如果不存在,返回-INF.
func (van *VanEmdeBoasTree) Prev(x int) int {
	return van.root.Prev(x)
}

// 返回大于等于i的最小元素.如果不存在,返回INF.
func (van *VanEmdeBoasTree) Next(x int) int {
	return van.root.Next(x)
}

func (van *VanEmdeBoasTree) Size() int {
	return van.size
}

// 如果没有元素,返回INF.
func (van *VanEmdeBoasTree) Min() int {
	return van.root.min
}

// 如果没有元素,返回-INF.
func (van *VanEmdeBoasTree) Max() int {
	return van.root.max
}

type VNode struct {
	min, max, dep int
	aux           *VNode
	son           map[int]*VNode
}

func NewVNode(dep int) *VNode {
	return &VNode{min: INF, max: -INF, dep: dep, son: map[int]*VNode{}}
}

func (v *VNode) Has(x int) bool {
	vMin, vMax := v.min, v.max
	if x == vMin || x == vMax {
		return true
	}
	vDep := v.dep
	if x < vMin || x > vMax || vDep == 0 {
		return false
	}
	i := x >> vDep
	soni, ok := v.son[i]
	if !ok {
		return false
	}
	return soni.Has(x - (i << vDep))
}

func (v *VNode) Insert(x int) {
	vMin, vMax := v.min, v.max
	if vMin > vMax {
		v.min, v.max = x, x
		return
	}
	if min, max := vMin, vMax; min == max {
		if x < min {
			v.min = x
			return
		} else if x > max {
			v.max = x
			return
		}
	}
	if x < vMin {
		x, v.min = vMin, x
	}
	if x > vMax {
		x, v.max = vMax, x
	}
	vDep := v.dep
	i := x >> vDep
	soni, ok := v.son[i]
	if !ok {
		soni = NewVNode(vDep >> 1)
		v.son[i] = soni
	}
	if soni.Empty() {
		if v.aux == nil {
			v.aux = NewVNode(vDep >> 1)
		}
		v.aux.Insert(i)
	}
	soni.Insert(x - (i << vDep))
}

func (v *VNode) Erase(x int) {
	vMin, vMax := v.min, v.max
	if vMin == x && vMax == x {
		v.min, v.max = INF, -INF
		return
	}

	vDep := v.dep
	aux := v.aux
	if x == v.min {
		if aux == nil || aux.Empty() {
			v.min = vMax
			return
		}
		auxMin := aux.min
		x = (auxMin << vDep) + v.son[auxMin].min
		v.min = x
	}
	if x == vMax {
		if aux == nil || aux.Empty() {
			v.max = vMin
			return
		}
		auxMax := aux.max
		x = (auxMax << vDep) + v.son[auxMax].max
		v.max = x
	}

	i := x >> vDep
	soni := v.son[i]
	soni.Erase(x - (i << vDep))
	if soni.Empty() {
		aux.Erase(i)
	}
}

func (v *VNode) Prev(x int) int {
	vMin := v.min
	if x < vMin {
		return -INF
	}
	vMax := v.max
	if x >= vMax {
		return vMax
	}
	vDep := v.dep
	i := x >> vDep
	hi := i << vDep
	lo := x - hi
	soni, ok := v.son[i]
	if ok && lo >= soni.min {
		return hi + soni.Prev(lo)
	}
	var y int
	if v.aux != nil && i > 0 {
		y = v.aux.Prev(i - 1)
	} else {
		y = -INF
	}
	if y == -INF {
		return vMin
	}
	return (y << vDep) + v.son[y].max
}

func (v *VNode) Next(x int) int {
	vMin := v.min
	if x <= vMin {
		return vMin
	}
	vMax := v.max
	if x > vMax {
		return INF
	}
	vDep := v.dep
	i := x >> vDep
	hi := i << vDep
	lo := x - hi
	soni, ok := v.son[i]
	if ok && lo <= soni.max {
		return hi + soni.Next(lo)
	}
	var y int
	if v.aux != nil {
		y = v.aux.Next(i + 1)
	} else {
		y = INF
	}
	if y == INF {
		return vMax
	}
	return (y << vDep) + v.son[y].min
}

func (v *VNode) Empty() bool {
	return v.min > v.max
}

File: 24_高级数据结构/珂朵莉树/VanEmdeBoasTree.ts
/* eslint-disable no-inner-declarations */
// Van Tree.
// 梵峨眉大悲寺树.

const INF = 2e15

/**
 * !建立一个元素范围为(-INF,INF)的VanEmdeBoasTree.
 * @warning 允许插入的元素范围为[0,INF).
 */
class VanEmdeBoasTree {
  private readonly _root: _VNode
  private _size = 0

  /**
   * @param depth 树的深度.默认为16.一般取16或32.
   */
  constructor(depth = 16) {
    this._root = new _VNode(depth)
  }

  has(x: number): boolean {
    return this._root.has(x)
  }

  /**
   * !0<=x<INF.
   */
  insert(x: number): boolean {
    if (x < 0) throw new Error('inserted value must be non-negative!')
    if (this.has(x)) return false
    this._size++
    this._root.insert(x)
    return true
  }

  erase(x: number): boolean {
    if (!this.has(x)) return false
    this._size--
    this._root.erase(x)
    return true
  }

  /**
   * 返回小于等于i的最大元素.如果不存在,返回-INF.
   */
  prev(x: number): number {
    return this._root.prev(x)
  }

  /**
   * 返回大于等于i的最小元素.如果不存在,返回INF.
   */
  next(x: number): number {
    return this._root.next(x)
  }

  /**
   * 遍历[start,end)区间内的元素.
   */
  enumerateRange(start: number, end: number, f: (v: number) => void): void {
    let x = start - 1
    while (true) {
      x = this.next(x + 1)
      if (x >= end) break
      f(x)
    }
  }

  toString(): string {
    const sb: string[] = []
    this.enumerateRange(-INF, INF, v => sb.push(v.toString()))
    return `VanEmdeBoasTree(${this.size}){${sb.join(', ')}}`
  }

  /**
   * 如果没有元素,返回INF.
   */
  get min(): number {
    return this._root.min
  }

  /**
   * 如果没有元素,返回-INF.
   */
  get max(): number {
    return this._root.max
  }

  get size(): number {
    return this._size
  }
}

class _VNode {
  dep: number
  min = INF
  max = -INF
  aux: _VNode | undefined = undefined
  son: Map<number, _VNode> = new Map()

  constructor(dep: number) {
    this.dep = dep
  }

  has(x: number): boolean {
    const { min: vMin, max: vMax, dep: vDep } = this
    if (x === vMin || x === vMax) return true
    if (!vDep || x < vMin || x > vMax) return false
    const i = x >>> vDep
    const soni = this.son.get(i)
    if (!soni) return false
    return soni.has(x - (i << vDep))
  }

  insert(x: number): void {
    const { min: vMin, max: vMax, dep: vDep } = this
    if (vMin > vMax) {
      this.min = x
      this.max = x
      return
    }
    if (vMin === vMax) {
      if (x < vMin) {
        this.min = x
        return
      }
      if (x > vMax) {
        this.max = x
        return
      }
    }
    if (x < vMin) {
      const tmp = x
      x = vMin
      this.min = tmp
    }
    if (x > vMax) {
      const tmp = x
      x = vMax
      this.max = tmp
    }
    const i = x >>> vDep
    let soni = this.son.get(i)
    if (!soni) {
      soni = new _VNode(vDep >>> 1)
      this.son.set(i, soni)
    }
    if (soni.empty()) {
      if (!this.aux) this.aux = new _VNode(vDep >>> 1)
      this.aux.insert(i)
    }
    soni.insert(x - (i << vDep))
  }

  erase(x: number): void {
    const { min: vMin, max: vMax, dep: vDep, aux: vAux } = this
    if (vMin === x && vMax === x) {
      this.min = INF
      this.max = -INF
      return
    }
    if (x === vMin) {
      if (!vAux || vAux.empty()) {
        this.min = vMax
        return
      }
      const auxMin = vAux.min
      x = (auxMin << vDep) + this.son.get(auxMin)!.min
      this.min = x
    }
    if (x === vMax) {
      if (!vAux || vAux.empty()) {
        this.max = vMin
        return
      }
      const auxMax = vAux.max
      x = (auxMax << vDep) + this.son.get(auxMax)!.max
      this.max = x
    }
    const i = x >>> vDep
    const soni = this.son.get(i)!
    soni.erase(x - (i << vDep))
    if (soni.empty()) vAux!.erase(i)
  }

  prev(x: number): number {
    const { min: vMin, max: vMax, dep: vDep } = this
    if (x < vMin) return -INF
    if (x >= vMax) return vMax
    const i = x >>> vDep
    const hi = i << vDep
    const lo = x - hi
    const soni = this.son.get(i)
    if (soni && lo >= soni.min) return hi + soni.prev(lo)
    let y = -INF
    if (this.aux && i > 0) y = this.aux.prev(i - 1)
    if (y === -INF) return vMin
    return (y << vDep) + this.son.get(y)!.max
  }

  next(x: number): number {
    const { min: vMin, max: vMax, dep: vDep } = this
    if (x <= vMin) return vMin
    if (x > vMax) return INF
    const i = x >>> vDep
    const hi = i << vDep
    const lo = x - hi
    const soni = this.son.get(i)
    if (soni && lo <= soni.max) return hi + soni.next(lo)
    let y = INF
    if (this.aux) y = this.aux.next(i + 1)
    if (y === INF) return vMax
    return (y << vDep) + this.son.get(y)!.min
  }

  empty(): boolean {
    return this.min > this.max
  }
}

export { VanEmdeBoasTree }

if (require.main === module) {
  const van = new VanEmdeBoasTree()
  console.log(van.min, van.max, van.size)
  van.insert(0)
  console.log(van.prev(0.5))
  console.log(van.prev(-1))
  console.log(van.prev(1))

  const n = 2e5
  const set2 = new VanEmdeBoasTree()
  console.time('VanEmdeBoasTree')
  for (let i = 0; i < n; i++) {
    set2.insert(i)
    set2.next(i)
    set2.prev(i)
    set2.has(i)
    set2.erase(i)
    set2.insert(i)
  }
  console.timeEnd('VanEmdeBoasTree') // 360ms

  testWithBruteForce()
  function testWithBruteForce(): void {
    class Mocker {
      private readonly _set = new Set<number>()
      insert(x: number): void {
        this._set.add(x)
      }
      erase(x: number): void {
        this._set.delete(x)
      }
      has(x: number): boolean {
        return this._set.has(x)
      }
      prev(x: number): number {
        let res = -INF
        for (const v of this._set) {
          if (v > x) continue
          if (v > res) res = v
        }
        return res
      }
      next(x: number): number {
        let res = INF
        for (const v of this._set) {
          if (v < x) continue
          if (v < res) res = v
        }
        return res
      }
    }

    const mocker = new Mocker()
    const set = new VanEmdeBoasTree()
    for (let i = 0; i < 4e4; i++) {
      const x = Math.floor(Math.random() * 1000)
      if (Math.random() < 0.5) {
        mocker.insert(x)
        set.insert(x)
      } else {
        mocker.erase(x - 500)
        set.erase(x - 500)
      }
      const x2 = Math.floor(Math.random() * 1000) - 500
      const prev1 = mocker.prev(x2)
      const prev2 = set.prev(x2)
      if (prev1 !== prev2) {
        console.log('prev1:', prev1, 'prev2:', prev2)
        throw new Error()
      }
      const next1 = mocker.next(x2)
      const next2 = set.next(x2)
      if (next1 !== next2) {
        console.log('next1:', next1, 'next2:', next2)
        throw new Error()
      }
    }
  }
  console.log('ok')
}
